### ISSUE TITLE
SPMC: No option to change resolution to 3840 × 2160

### ISSUE THREAD

**Issue by ianwrd4** [koying/SPMC#888]
_Hi
Im using the Nvidia Shield which currently does not include automatic switching to the  native resolution of the source. Therefore when I want to watch 4K content I will first re-set the Shiled's resolution from 1080p to 4K/60fps. (I do it this way as I usually watch HD & the TV does a much better conversion from 1080p to 4K than the Shield FYI LG 65B6 OLED)
Now when I go to SPMC to change the resolution settings to 3840 × 2160 it only gives me the option for HD 1920 × 1080, no option for 3840 × 2160? Btw KODI 17.1 has the same issue. This did not happen with previous versions of SPMC, as for Kodi I cant comment!
However if I switch my whole system off and then restart the system TV first - TV > AVR> Shield then the 3840 × 2160 resolution option is available! I have checked also on the Minix U9 and the issue is the same when it comes to the resolution settings.
Is there any way around this or do I just need to remember to switch on my TV device first?

**Comment by barenaked** [comment]
That's an issue with the shield, you have to restart for system changes to become fully applied.

**Comment by ianwrd4** [comment]
If u read my post correctly you will see its not just the SHIELD, as I tried the same on the new MINI U9 and it was the same no option for UHD? This is not ANDROID box related, it has something to do with KODI & SPMC. To confirm this there was no issue with previous versions. Yes the SHIELD has an issue regarding resolution, by not having an option for automatic resolution settings based on input resolution but thats another issue!

**Comment by wrxtasy** [comment]
But what's the point when 720/1080p <<-->> 2160p(4K) Auto resolution switching will not work anyway ?

Even if the option was there it's useless.

**There is no App - Auto Resolution switching implemented in Firmware** on 99% of Android Boxes apart from the 4K - FireTV2:
https://developer.amazon.com/public/solutions/devices/fire-tv/docs/fire-tv-4k-ultra-hd#playing-4k-ultra-hd-videos-on-fire-tv

**Comment by ianwrd4** [comment]
And the MINIX U9 I have one here and it works perfectly so cant be that hard to do, especially for the price we are paying for the SHIELD. Anyway back to my original topic of why doesnt SPMC/KODI recognise a 4K signal and offer the correct resolution choice in the system>resolution settings . . .

**Comment by MrMC** [comment]
Put shield display on auto, then display switching will show up. That's the entire point of adding it. 

Note: FireTV1's can display switch as well but they do not support 4k.

**Comment by ianwrd4** [comment]
MrMC
What am I missing here > 
The SHIELD doesn't have an "AUTO" setting in resolution, OR at least mine doesn't? Would you be so kind as to explain further for the not so gifted?
Thanks in advance

![20170503_073000](https://cloud.githubusercontent.com/assets/28243191/25644820/1b48daee-2fd3-11e7-9971-10bfa8ef0644.jpg)

**Comment by MrMC** [comment]
sorry, confusing shield with firetv :) pick 4k 59.940 (recommended).

**Comment by ianwrd4** [comment]
Thanks for clarification I was looking everywhere for that auto selection lol :-) The issue I have with selecting 4K/59.940 is the SHIELD will then up convert all 720/1080p material to 4K, and it does a bad job of that compared to my LG B6 OLED. So I prefer to have it set at 1080p/60 so that the TV does the up scaling to 4K and not the SHIELD, the picture is far superior this way for 1080p/720p material. 
It would be nice if the SHIELD had the AUTO selection!

**Comment by MrMC** [comment]
1) set Shield to 4K.
2) set SPMC/Kodi/MrMC GUI display size to 1080p, that changes display for app to that size. 

If you home out to shield home, it will switch back to 4K, go into SPMC/Kodi/MrMC, it will switch to what you set in display size. That is default behavior for android display switching. FireTV does it too.

Enable display switching. Now, when you play 1080p/24, it switches to that and back back when you stop. Play 4K, it will switch to 4K.

MrMC can do 2), Pretty sure SPMC can do it too. Kodi... not sure.

**Comment by ianwrd4** [comment]
Thanks very much I will need to have a play arround!

**Comment by CLFNole** [comment]
Why is it recommended to pick 4K 59.940Hz instead of 4K 60Hz?

**Comment by wesk05** [comment]
>If you home out to shield home, it will switch back to 4K, go into SPMC/Kodi/MrMC, it will switch to what you set in display size. That is default behavior for android display switching. FireTV does it too.
Enable display switching. Now, when you play 1080p/24, it switches to that and back back when you stop. Play 4K, it will switch to 4K.
MrMC can do 2), Pretty sure SPMC can do it too. Kodi... not sure.

I don't see this working on my Shield 2017 with MrMC or SPMC. The resolution is whatever is set in Shield's setting. I don't see it switching to 1080p from 4K or the other way.
Where is this display switching setting in MrMC? I only  see refresh rate switching option.

**Comment by MrMC** [comment]
Same place it's always been, settings -> system -> video output -> Resolution.

But you are correct, Shield always outputs 4K. It's FireTV2 that allows this.

**Comment by wesk05** [comment]
@MrMC : Oh! you were referring to the resolution options (1080p, 4K). When you said "enable display switching" I misunderstood that as a setting to enable display switching just like "adjust display refresh rate".

**Comment by ianwrd4** [comment]
Im totaly confused now lol :-)

**Comment by wrxtasy** [comment]
I will repeat this again as MrMC introduced Mass confusion ;-)

**There is no App - Auto Resolution switching implemented in Firmware on 99% of Android Boxes apart from the 4K - FireTV2:**

https://developer.amazon.com/public/solutions/devices/fire-tv/docs/fire-tv-4k-ultra-hd#playing-4k-ultra-hd-videos-on-fire-tv

Do not confuse this with Auto Refresh Switching - Adjust display refresh rate option in Kodi.

So the solution - buy a 4K FireTV2 OR buy a AML S905 and run LibreELEC Kodi Krypton on it - as these have Auto 1080p <<-->> 2160p Resolution switching.

Both of these are superior 4K Kodi solutions for viewing 720/1080p/2160p content to what the Shield currently has. The FireTV2 will be limited to 2160p @ a Max 30Hz.

You can then let the TV with its superior hardware, upscale 720/1080p > 2160p (4K) video. Only when actually viewing 4K content will the TV Auto switch to a 4K Resolution.

You may get acceptable results when configuring SPMC like [THIS:](https://github.com/koying/SPMC/wiki/Recommended-video-settings-for-SPMC%2C-NVIDIA-Shield-TV-and-UHD-%284K%29-TVs-with-4K-GUI)

And it would also be wise to [READ THIS](http://forum.kodi.tv/showthread.php?tid=252916&pid=2581963#pid2581963) recent Kodi upscaling post, regarding the Shield.

In reality what needs to be done is a whole bunch of aggrieved 4K nVIDIA Shield owners need to nag nVIDIA constantly over on their forums - to implement App Auto Resolution switching.
Create a massive thread to get their attention.

**Comment by ianwrd4** [comment]
Thanks a lot for taking the time to repeat yourself much appreciated, after all of that I think the actual process of switching over setting every time I watch 4K isn't to much bother after all. I will take your instructions on board snd see how I end up.
Thanks again
IW

**Comment by MrMC** [comment]
:) Some call me Mr Massive Confusion

**Comment by wrxtasy** [comment]
Ha ha - very clever !

**Comment by jgeee** [comment]
@wrxtasy while I understand your point about libreelec for resolution switching, it might be a bit misleading for some 4k users depending on use cases.

I'm a 4k aml s905x user who just switched from libreelec to android because the LE aml builds are incapable of 10 bit output or hdr...and the 8 bit conversion suffers from very bad color banding (though not as bad on jarvis builds). This is something most 4k users will care about.

Since most of these devices are capable of doing 720 output without hw accel, the better compromise might be to patch spmc to support setting no hw accel on 720 or lower, and then just dealing with the poor upscaling (which isn't as noticeable) with 1080 content.

A lot of these amlogic android devices are also still trying to deinterlace everything, so setting the bypass interface helps with upscaling pq a bit.

Edit: I forgot about another bug with the aml le builds: changing colorspace isn't supported unlike on android.

Ultimately there are many cons to aml libreelec that 4k users are more likely to care about, and each new release (jarvis to krypton to le master) seems to introduce new regressions (dithering regressions and 4:4:4 support removal).

**Comment by wrxtasy** [comment]
Yes AML LibreELEC does not handle HDR10 properly or have an option to change to Deep color mode and everything is dithered to an 8bit color output, but that was not the point of this whole thread - it was about resolution switching - which AML LibreELEC does seamlessly.

The reality is there is a no One Box solution to meet all Kodi - 1080p / 4K HDR10 / reliable deinterlacing - Android Apps / Auto resolution switching / high quality upscaling solutions. Trying to do it seamlessly, in a hands off manner on the one media player box is commonly where the wheels fall off ALL the media player wagons...

Now off topic....

The other problem is the Shield actually has noticeable 10/12bit Kodi/SPMC color mapping output problems as detailed from [this post onwards](http://forum.kodi.tv/showthread.php?tid=304226&pid=2585277#pid2585277)

In reality 4K HDR & 4K resolutions would be easier to use with actual Apps on a Smart 4K HDR TV - let the TV handle all the 720/1080p > 4K Android video streaming Apps upscaling and Resolution switching. The color mapping will likely be correct then too.

OR investigate a Dual boot OS - external media player solution, and leave Android in a 4K resolution for 4K HDR media content playback only.
